### PR TITLE
remote: add multicast discovery w/ encrypted announce

### PR DESCRIPTION
**Pull request by dpwiz**

### CODE REVIEW

File: src/Simplex/RemoteControl/Client.hs [hunk 1]
```diff
@@ -76,6 +78,9 @@ xrcpBlockSize = 16384
 helloBlockSize :: Int
 helloBlockSize = 12288
 
+announceBlockSize :: Int
+announceBlockSize = 1024 -- TODO: put a real number
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Client.hs:82]: what's the number? or 1024 is ok?

File: src/Simplex/RemoteControl/Client.hs [hunk 2]
```diff
@@ -89,33 +94,40 @@ data RCHostClient = RCHostClient
...
+  c@RCHClient_ {startedPort, announcer} <- liftIO mkClient
   hostKeys <- liftIO genHostKeys
   action <- runClient c r hostKeys `putRCError` r
   -- wait for the port to make invitation
   -- TODO can't we actually find to which interface the server got connected to get host there?
   portNum <- atomically $ readTMVar startedPort
-  signedInv <- maybe (throwError RCETLSStartFailed) (liftIO . mkInvitation hostKeys host) portNum
+  signedInv@RCSignedInvitation{invitation} <- maybe (throwError RCETLSStartFailed) (liftIO . mkInvitation hostKeys host) portNum
+  when multicast $ case knownHost of
+    Nothing -> fail "oops, must have known host for multicast"
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Client.hs:116]: this should be some kind of "prohibited" error, not hard fail

File: src/Simplex/RemoteControl/Invitation.hs [hunk 3]
```diff
@@ -174,3 +157,67 @@ requiredP q k f = maybe (fail $ "missing " <> show k) (either fail pure . f) $ l
...
+sessionAddressJSON = LB.toStrict . J.encode . J.toJSON
+
+signInviteJSON :: C.PrivateKey C.Ed25519 -> C.PrivateKey C.Ed25519 -> RCInvitation -> ByteString
+signInviteJSON sKey idKey invitation = idSigned
+  where
+    sessionAddress = sessionAddressJSON invitation
+
+    ssig :: C.Signature 'C.Ed25519
+    ssig =
+      case C.sign (C.APrivateSignKey C.SEd25519 sKey) sessionAddress of
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Invitation.hs:171]: you should use type-specific sign'

File: src/Simplex/RemoteControl/Invitation.hs [hunk 4]
```diff
@@ -174,3 +157,67 @@ requiredP q k f = maybe (fail $ "missing " <> show k) (either fail pure . f) $ l
...
+        C.ASignature C.SEd25519 s -> s
+        _ -> error "signing with ed25519"
+    sSigned = sessionAddress <> C.signatureBytes ssig
+
+    idsig :: C.Signature 'C.Ed25519
+    idsig =
+      case C.sign (C.APrivateSignKey C.SEd25519 idKey) sSigned of
+        C.ASignature C.SEd25519 s -> s
+        _ -> error "signing with ed25519"
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Invitation.hs:180]: same here

File: src/Simplex/RemoteControl/Invitation.hs [hunk 5]
```diff
@@ -174,3 +157,67 @@ requiredP q k f = maybe (fail $ "missing " <> show k) (either fail pure . f) $ l
...
+    idSigned = sSigned <> C.signatureBytes idsig
+
+instance Encoding RCSignedInvitation where
+  smpEncode RCSignedInvitation {invitation, ssig, idsig} = sessionAddressJSON invitation <> C.signatureBytes ssig <> C.signatureBytes idsig
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Invitation.hs:184]: this is brittle and should instead have length prefixes, so it should be something like smpEncode (sessionAddressJSON invitation, ssig, idsig)

Signatures know how to encode.

Then your decoder will be as simple:

(json, ssig, idsig) <- smpP

File: src/Simplex/RemoteControl/Types.hs [hunk 6]
```diff
@@ -31,11 +31,13 @@ data RCErrorType
   = RCEInternal {internalErr :: String}
   | RCEIdentity
   | RCENoLocalAddress
+  | RCENotDiscovered
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Types.hs:34]:
```suggestion
  | RCETimeout
```

File: src/Simplex/RemoteControl/Types.hs [hunk 7]
```diff
@@ -31,11 +31,13 @@ data RCErrorType
...
   | RCETLSStartFailed
   | RCEException {exception :: String}
   | RCECtrlAuth
   | RCECtrlNotFound
   | RCECtrlError {ctrlErr :: String}
+  | RCEAddress
```
Comment (epoberezkin) [inline src/Simplex/RemoteControl/Types.hs:40]: not sure we need a special error - it's just invalid announcement?

File: src/Simplex/RemoteControl/Types.hs (hunk 6 above)
Comment (dpwiz) [reply src/Simplex/RemoteControl/Types.hs:34]: Makes RCETimeout in chat ambiguous

File: src/Simplex/RemoteControl/Client.hs (hunk 1 above)
Comment (dpwiz) [reply src/Simplex/RemoteControl/Client.hs:82]: something under MTU while covering announcement json,
1024 should be ok.

**Pull request merged**